native/common/jp_array.cpp

```diff
@@ -128,8 +123,46 @@ JPPyObject JPArray::getItem(jsize ndx)
 
 	if (ndx >= m_Length || ndx < 0)
 	{
-		JP_RAISE_INDEX_ERROR("array index out of bounds");
+		JP_RAISE(PyExc_IndexError, "array index out of bounds");
 	}
 
-	return compType->getArrayItem(frame, m_Object.get(), ndx);
+	return compType->getArrayItem(frame, m_Object.get(), m_Start + ndx * m_Step);
+}
+
+jarray JPArray::clone(JPJavaFrame& frame, PyObject* obj)
+{
+	JPValue value = m_Class->newInstance(frame, m_Length);
+	JPClass* compType = m_Class->getComponentType();
+	jarray out = (jarray) value.getValue().l;
+	compType->setArrayRange(frame, out, 0, m_Length, 1, obj);
+	return out;
+}
+
+JPArrayView::JPArrayView(JPArray* array)
+{
+	this->array = array;
+	refcount = 0;
+	buffer.obj = NULL;
+	buffer.ndim = 1;
+	buffer.suboffsets = NULL;
+	array->getClass()->getComponentType()->getView(*this);
+	strides[0] = buffer.itemsize * array->m_Step;
+	shape[0] = array->m_Length;
+	buffer.buf = (char*) memory + buffer.itemsize * array->m_Start;
+	buffer.len = array->m_Length * buffer.itemsize;
+	buffer.shape = shape;
+	buffer.strides = strides;
+	buffer.readonly = 1;
+}
+
+void JPArrayView::reference()
+{
+	refcount++;
+}
+
+bool JPArrayView::unreference()
+{
+	refcount--;
+	array->getClass()->getComponentType()->releaseView(*this, refcount == 0);
```

Comment:
shouldn't the refcount check happen here, rathen than demangling the type and doing a null-op, if refcount is still positive?

Reply:
The code was supposed to call a commit with each dereference.  The refcounts represent the number of active buffers.  There is a difficulty that java does not allow write through, thus no changes actually commit until the view is deleted.  For most transfers this did not matter.  But memoyview which holds it open made a huge deal.   Thus the best i could achieve was to try to commit with each close hear.  

I personally think this is too awkward to be usable as the user can only control if the only have one open.  Someone may try to tell numpy to use an array on the memorybuf which would not write through until the last copy dies.  

The alternative is to force read only.  But as most buffer transactions request read and write even if the only need read this ended up not working well.

Reply:
The old impl commited after each write request (which implies releasing the actual jni view, which may cause the jvm array to be freed). So your pattern protect against this. It also means that we can leak JVM memory, if the Python view is somehow never released. But seems much more stable. I would not enforce read-only, it is too restrictive.

Reply:
The previous implementation was a copy once on request solution so by definition it was write through.  It did have the downside that readonly was not enforced thus you could potentially alter the array copy using the double slice operator but it was never reflected in Java so it leaves the impression that work was done that wasn't.

``` 
ja=JArray(JInt)([1,2,3,4)
b = ja[:]
b[1] = 5
print(ja)  <== produces [1,2,3,4]
```

The new implementation writes through because the slice is actually a view.  But it wont permit writing through of the memory view because we can't pin the array.  Hence you may be confused between the writing of the slice view and writing of the byte buffer view than backs it.  We can support writing the slice just fine now, but byte buffer which requires creating a memoryview, accessing the memory, then deleting the view is problematic.

Reply:
Sorry I edited the last comment... should have read "read only" not "write through"